services/core/MasterDriverAgent/master_driver/agent.py

```diff
@@ -400,12 +427,166 @@ def heart_beat(self):
+        pattern = pattern.lower()
+        config_update = False
+
+        # If pattern exactly matches
+        if pattern in self._override_patterns:
+            self._override_patterns.discard(pattern)
+            # Cancel any pending override events
+            self._cancel_override_events(pattern)
+            config_update = True
+        else:
+            _log.debug("Pattern did not match!")
+
+        self._override_devices.clear()
+        patterns = dict()
+        #Build override devices list again
+        for pat in self._override_patterns:
+            for device in self.instances.keys():
+                device = device.lower()
+                if fnmatch.fnmatch(device, pat):
+                    self._override_devices.add(device)
+            if config_update:
+                if self._override_interval_events[pat] is None:
+                    patterns[pat] = None
+                else:
+                    evt, end_time = self._override_interval_events[pat]
+                    patterns[pat] = end_time.isoformat()
+        if config_update:
+            self.vip.config.set("override_patterns", jsonapi.dumps(patterns))
+
+    def _update_override_interval(self, interval, pattern, indefinite_duration=False):
+        if indefinite_duration:
+            if pattern in self._override_interval_events.keys():
+                #If set as indifinite, do nothing
+                if self._override_interval_events[pattern] == None:
+                    return False
+                else:
+                    #Cancel the event
+                    evt = self._override_interval_events.pop(pattern)
+                    evt[0].cancel()
+            self._override_interval_events[pattern] = None
+            return True
+        else:
+            override_start = utils.get_aware_utc_now()
+            override_end = override_start + timedelta(seconds=interval)
+            if pattern in self._override_interval_events.keys():
+                evt = self._override_interval_events[pattern]
+                #If event is indefinite or greater than new end time, do nothing
+                if evt is None or override_end < evt[1]:
+                    return False
+                else:
+                    evt = self._override_interval_events.pop(pattern)
+                    evt[0].cancel()
+            event = self.core.schedule(override_end, self._cancel_override, pattern)
+            self._override_interval_events[pattern] = (event, override_end)
+            return True
+
+    def _cancel_override_events(self, pattern):
+        if pattern in self._override_interval_events:
+            # Cancel the override cancellation timer event
+            evt = self._override_interval_events.pop(pattern, None)
+            if evt is not None:
+                evt[0].cancel()
+
+    def _cancel_override(self, topic):
+        self._set_override_off(topic)
+
+    def _update_override_state(self, device, state):
+        device = device.lower()
+
```

Comment:
In the case where state == 'remove' we should not do any comparisons to the override patterns and just remove the device from self._override_devices if it's in there.